login/profile.go

```diff
@@ -136,46 +156,150 @@ func (userProfileClient *KeycloakUserProfileClient) Create(ctx context.Context,
 		"user_url":          createdUserURLString,
-	}, "Successfully created Keycloak user user")
+	}, "Successfully created Keycloak user")
+
+	return &createdUserURLString, false, nil
+}
+
+func (userProfileClient *KeycloakUserProfileClient) updateAsAdmin(ctx context.Context, keycloakUserRequest *KeytcloakUserRequest, protectedAccessToken string, keycloakAdminUserAPIURL string) (*string, error) {
+	user, err := userProfileClient.loadUser(ctx, *keycloakUserRequest.Username, protectedAccessToken, keycloakAdminUserAPIURL)
+	if err != nil {
+		return nil, err
+	}
+	if user == nil {
+		log.Error(ctx, map[string]interface{}{
+			"keycloak_user_profile_url": keycloakAdminUserAPIURL,
+			"email":                     *keycloakUserRequest.Email,
+		}, "Unable to update Keycloak user because user not found")
+		return nil, errs.New("unable to update Keycloak user because user not found")
+	}
+	body, err := json.Marshal(keycloakUserRequest)
+	if err != nil {
+		return nil, errors.NewInternalError(ctx, err)
+	}
+	userURL := keycloakAdminUserAPIURL + "/" + *user.ID
+	req, err := http.NewRequest("PUT", userURL, bytes.NewReader(body))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", "Bearer "+protectedAccessToken)
+	req.Header.Add("Content-Type", "application/json")
+
+	resp, err := userProfileClient.client.Do(req)
 
-	return &createdUserURLString, nil
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"keycloak_user_profile_url": keycloakAdminUserAPIURL,
+			"email":                     *keycloakUserRequest.Email,
+			"err":                       err,
+		}, "Unable to update Keycloak user")
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	bodyString := rest.ReadBody(resp.Body)
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		log.Error(ctx, map[string]interface{}{
+			"response_status":           resp.Status,
+			"response_body":             bodyString,
+			"keycloak_user_profile_url": keycloakAdminUserAPIURL,
+			"email":                     *keycloakUserRequest.Email,
+		}, "Unable to update Keycloak user")
+		return nil, errs.Errorf("received a non-2xx response %s while creating keycloak user:  %s", resp.Status, keycloakAdminUserAPIURL)
+	}
+	log.Info(ctx, map[string]interface{}{
+		"response_status":           resp.Status,
+		"response_body":             bodyString,
+		"keycloak_user_profile_url": keycloakAdminUserAPIURL,
+		"email":                     *keycloakUserRequest.Email,
+	}, "Successfully updated Keycloak user")
+
+	return &userURL, nil
+}
+
+// loadUser search for a user by username. Return nil if no user found.
+func (userProfileClient *KeycloakUserProfileClient) loadUser(ctx context.Context, username string, protectedAccessToken string, keycloakAdminUserAPIURL string) (*KeycloakUserProfile, error) {
+	kcURL, err := rest.AddParams(keycloakAdminUserAPIURL, map[string]string{
+		"username": username,
+		"first":    "0",
+		"max":      "500", // TODO we need to handle big user lists better
```

Comment:
umm. why do we need this ? Does this imply that we search upto the first 500 results?

Reply:
Yeah. I thought it's better to fail updating a user if there are too many matches than fail because of overloading KC and Auth if the result is too big. But I don't like this either :(